File: trunk/managed/Warp3D/Warp3DVB/warp_Texture.vb
' Created by: X
' http://www.createdbyx.com/
' ----------------------
' C# to VB.NET conversion of this code was done by hand and with the assistance of the 
' Convert C# to VB .NET application availible at http://www.kamalpatel.net/ConvertCSharp2VB.aspx
' ConvertCSharp2VB utility was developed by: Kamal Patel (http://www.KamalPatel.net) 
' ----------------------
' Thanks to Alan Simes ( http://alansimes.blogdns.net/ ) for creating the Warp3D library.
' http://alansimes.blogdns.net/cs/files/default.aspx
' ----------------------

Imports System
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.Net

Namespace Rednettle.Warp3D
    '/ <summary>
    '/ Summary description for warp_Texture.
    '/ </summary>

    Public Class warp_Texture
        Public width As Integer
        Public height As Integer
        Public bitWidth As Integer
        Public bitHeight As Integer
        Public pixel() As Integer
        Public path As String = Nothing

        Public Shared ALPHA As Integer = &HFF000000  ' alpha mask

        Public Sub New(ByVal w As Integer, ByVal h As Integer)
            height = h
            width = w
            pixel = New Integer(w * h) {}
            cls()
        End Sub

        Public Sub New(ByVal w As Integer, ByVal h As Integer, ByVal data() As Integer)
            height = h
            width = w
            pixel = New Integer(width * height) {}

            System.Array.Copy(data, pixel, pixel.Length)
        End Sub

        Public Sub New(ByVal path As String)
            Dim map As Bitmap

            If path.StartsWith("http") Then
                Dim webrq As WebRequest = WebRequest.Create(path)
                map = CType(Bitmap.FromStream(webrq.GetResponse().GetResponseStream()), Bitmap)
            Else
                map = New Bitmap(path, False)
            End If

            loadTexture(map)
        End Sub


        Public Sub resize()
            Dim log2inv As Double = 1 / Math.Log(2)

            '    int w = ( int )Math.Pow( 2, bitWidth = ( int )Math.Ceiling( ( Math.Log( width ) * log2inv ) ) );
            '   int h = ( int )Math.Pow( 2, bitHeight = ( int )Math.Ceiling( ( Math.Log( height ) * log2inv ) ) );
            bitWidth = CInt(Math.Ceiling(Math.Log(width) * log2inv))
            bitHeight = CInt(Math.Ceiling(Math.Log(height) * log2inv))
            Dim w As Integer = CInt(Math.Pow(2, bitWidth))
            Dim h As Integer = CInt(Math.Pow(2, bitHeight))

            If Not (w = width AndAlso h = height) Then resize(w, h)
        End Sub


        Public Sub resize(ByVal w As Integer, ByVal h As Integer)
            '   System.Console.WriteLine("warp_Texture| resize :" + w + "," + h)
            setSize(w, h)
        End Sub

        Public Sub cls()
            warp_Math.clearBuffer(pixel, 0)
        End Sub

        Public Function toAverage() As warp_Texture
            Dim i As Integer
            For i = width * height - 1 To 0 Step -1
                pixel(i) = warp_Color.getAverage(pixel(i))
            Next

            Return Me
        End Function

        Public Function toGray() As warp_Texture
            Dim i As Integer
            For i = width * height - 1 To 0 Step -1
                pixel(i) = warp_Color.getGray(pixel(i))
            Next

            Return Me
        End Function

        Public Function valToGray() As warp_Texture
            Dim intensity As Integer
            Dim i As Integer
            For i = width * height - 1 To 0 Step -1
                intensity = warp_Math.crop(pixel(i), 0, 255)
                pixel(i) = warp_Color.getColor(intensity, intensity, intensity)
            Next

            Return Me
        End Function

        Public Function colorize(ByVal pal() As Integer) As warp_Texture
            Dim range As Integer = pal.GetLength(0) - 1
            Dim i As Integer
            For i = width * height - 1 To 0 Step -1
                pixel(i) = pal(warp_Math.crop(pixel(i), 0, range))
            Next

            Return Me
        End Function

        Private Sub loadTexture(ByVal map As Bitmap)
            width = map.Width
            height = map.Height

            pixel = New Integer((width * height) - 1) {}

            'Dim scanline As Integer = bmData.Stride
            'Dim Scan0 As System.IntPtr = bmData.Scan0
            Dim pIndex As Integer = 0
            '	Dim *p As Byte = (Byte *)(void *)Scan0 
            'Dim nOffset As Integer = bmData.Stride - map.Width * 3
            Dim nPixel As Integer = 0
            Dim p((width * height * 3) - 1) As Byte

            Dim bmData As BitmapData = map.LockBits(New Rectangle(0, 0, map.Width, map.Height), ImageLockMode.ReadOnly, PixelFormat.Format24bppRgb)
            System.Runtime.InteropServices.Marshal.Copy(bmData.Scan0, p, 0, p.Length)
            map.UnlockBits(bmData)

            For i As Integer = 0 To map.Height - 1 'Step i + 1
                For j As Integer = 0 To map.Width - 1 'Step j + 1
                    ' Dim Col As Color
                    ' Col = map.GetPixel(j, i)
                    Dim blue As Integer = p(pIndex)   'Col.B 
                    Dim green As Integer = p(pIndex + 1)  'Col.G
                    Dim red As Integer = p(pIndex + 2)        'Col.R 
                    pixel(nPixel) = ALPHA Or red << 16 Or green << 8 Or blue
                    nPixel += 1

                    pIndex += 3
                Next

                ' pIndex += nOffset
            Next

            resize()
        End Sub

        'Private Sub loadTexture(ByVal map As Bitmap)
        '    width = map.Width
        '    height = map.Height

        '    pixel = New Integer((width * height) - 1) {}

        '    Dim bmData As BitmapData = map.LockBits(New Rectangle(0, 0, map.Width, map.Height), ImageLockMode.ReadWrite, PixelFormat.Format24bppRgb)
        '    Dim scanline As Integer = bmData.Stride
        '    Dim Scan0 As System.IntPtr = bmData.Scan0
        '    Dim pIndex As Integer = 0
        '    '	Dim *p As Byte = (Byte *)(void *)Scan0 
        '    Dim nOffset As Integer = bmData.Stride - map.Width * 3
        '    Dim nPixel As Integer = 0
        '    Dim p(width * height * 3) As Byte

        '    System.Runtime.InteropServices.Marshal.Copy(Scan0, p, nOffset, p.Length)

        '    For i As Integer = 0 To map.Height - 1 'Step i + 1
        '        For j As Integer = 0 To map.Width - 1 'Step j + 1
        '            Dim blue As Integer = p(pIndex)
        '            Dim green As Integer = p(pIndex + 1)
        '            Dim red As Integer = p(pIndex + 2)
        '            pixel(nPixel) = ALPHA Or red << 16 Or green << 8 Or blue
        '            nPixel += 1

        '            pIndex += 3
        '        Next

        '        pIndex += nOffset
        '    Next

        '    map.UnlockBits(bmData)
        '    resize()
        'End Sub
 
        Private Sub setSize(ByVal w As Integer, ByVal h As Integer)
            Dim offset As Integer = w * h
            Dim offset2 As Integer
            If w * h <> 0 Then
                Dim Newpixels() As Integer = New Integer(w * h) {}
                Dim j As Integer
                For j = h - 1 To 0 Step -1
                    offset -= w
                    offset2 = CInt((j * height / h) * width)
                    Dim i As Integer
                    For i = w - 1 To 0 Step -1
                        Newpixels(i + offset) = pixel(CInt((i * width / w) + offset2))
                    Next
                Next

                width = w
                height = h
                pixel = Newpixels
            End If
        End Sub

        Private Function inrange(ByVal a As Integer, ByVal b As Integer, ByVal c As Integer) As Boolean
            Return (a >= b) And (a < c)
        End Function

        Public Function getClone() As warp_Texture
            Dim t As warp_Texture = New warp_Texture(width, height)
            warp_Math.copyBuffer(pixel, t.pixel)
            Return t
        End Function

    End Class
End Namespace

File: trunk/managed/Warp3Dmod/warp_Texture.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.Net;

namespace Warp3D
{
    /// <summary>
    /// Summary description for warp_Texture.
    /// </summary>

    public class warp_Texture
    {
        public int width;
        public int height;
        public int bitWidth;
        public int bitHeight;
        public int[] pixel;
        public String path = null;
        public int averageColor;
        public bool hasAlpha;
        public bool opaque;
        public bool hasmips;
        public int[][] mips;
        public int[] mipsBitWidth;
        public int[] mipsBitHeight;
        public int[] mipstw;
        public int[] mipsth;
        public int maxmips;

        public static int ALPHA = unchecked((int)0xFF000000); // alpha mask

        public warp_Texture(int w, int h)
        {
            height = h;
            width = w;
            pixel = new int[w * h];
            cls();
        }

        public warp_Texture(int w, int h, int[] data)
        {
            height = h;
            width = w;
            pixel = new int[width * height];

            System.Array.Copy(data, pixel, pixel.Length);
        }

        public warp_Texture(Bitmap map, int maxBitSize = -1)
        {
            loadTexture(map, maxBitSize);
        }

        public warp_Texture(string path, int maxBitSize = -1)
        {
            using(Bitmap map = new Bitmap(path, false))
                loadTexture(map, maxBitSize);
        }

        public void resize()
        {
            double log2inv = 1 / Math.Log(2);

            int w = (int)Math.Pow(2, bitWidth = (int)Math.Ceiling((Math.Log(width) * log2inv)));
            int h = (int)Math.Pow(2, bitHeight = (int)Math.Ceiling((Math.Log(height) * log2inv)));

            if(!(w == width && h == height))
                resize(w, h);
        }

        public void resize(int w, int h)
        {
            setSize(w, h);
        }

        public void cls()
        {
            warp_Math.clearBuffer(pixel, 0);
        }

        public warp_Texture toAverage()
        {
            for(int i = width * height - 1; i >= 0; i--)
                pixel[i] = warp_Color.getAverage(pixel[i]);

            return this;
        }

        public warp_Texture toGray()
        {
            for(int i = width * height - 1; i >= 0; i--)
                pixel[i] = warp_Color.getGray(pixel[i]);

            return this;
        }

        public warp_Texture valToGray()
        {
            int intensity;
            for(int i = width * height - 1; i >= 0; i--)
            {
                intensity = warp_Math.crop(pixel[i], 0, 255);
                pixel[i] = warp_Color.getColor(intensity, intensity, intensity);
            }

            return this;
        }

        public warp_Texture colorize(int[] pal)
        {
            int range = pal.GetLength(0) - 1;
            for(int i = width * height - 1; i >= 0; i--)
                pixel[i] = pal[warp_Math.crop(pixel[i], 0, range)];

            return this;
        }

        private void loadTexture(Bitmap pmap, int maxBitSize = -1)
        {
            Bitmap map = pmap;
            bool disposemap = false;

            opaque = true;
            width = map.Width;
            height = map.Height;

            PixelFormat format = pmap.PixelFormat;
            PixelFormat targetformat;

            switch(format)
            {
                case PixelFormat.Format16bppArgb1555:
                case PixelFormat.Format32bppArgb:
                case PixelFormat.Format32bppPArgb:
                case PixelFormat.Format64bppArgb:
                case PixelFormat.Format64bppPArgb:
                    targetformat = PixelFormat.Format32bppArgb;
                    break;
                default:
                    targetformat = PixelFormat.Format24bppRgb;
                    break;
            }

           // make sure sizes are power of 2
            const double log2inv = 1.4426950408889634073599246810019;
            bitWidth = (int)Math.Ceiling((Math.Log(width) * log2inv));
            bitHeight = (int)Math.Ceiling((Math.Log(height) * log2inv));

            if(maxBitSize > 3)
            {
                if (bitWidth > maxBitSize)
                    bitWidth = maxBitSize;
                if (bitHeight > maxBitSize)
                    bitHeight = maxBitSize;
            }

            int w = (int)Math.Pow(2, bitWidth);
            int h = (int)Math.Pow(2, bitHeight);

            if(width != w || height != h || format != targetformat)
            {
                map = ResizeBitmap(pmap, w, h, targetformat);
                width = w;
                height = h;
                disposemap = true;
            }

            ulong avgR = 0;
            ulong avgG = 0;
            ulong avgB = 0;
            ulong avgA = 0;

            byte blue;
            byte green;
            byte red;
            byte alpha;
            
            pixel = new int[width * height];
            hasAlpha = targetformat == PixelFormat.Format32bppArgb;

            BitmapData bmData = map.LockBits(new Rectangle(0, 0, map.Width, map.Height),
                            ImageLockMode.ReadOnly, targetformat);
            unsafe
            {
                byte* p = (byte*)bmData.Scan0;
                fixed(int* px = pixel)
                {
                    int nPixel = 0;
                    if(hasAlpha)
                    {
                        int nOffset = bmData.Stride - width * 4;
                        for(int i = 0; i < height; i++)
                        {
                            for(int j = 0; j < width; j++)
                            {
                                blue = *(p++);
                                avgB += blue; 
                                green = *(p++);
                                avgG += green; 
                                red = *(p++);
                                avgR += red;
                                alpha = *(p++);
                                avgA += alpha;

                                if(opaque && alpha != 0xff)
                                    opaque = false;
                                px[nPixel++] = alpha << 24 | red << 16 | green << 8 | blue;
                            }
                            p += nOffset;
                        }
                    }
                    else
                    {
                        int nOffset = bmData.Stride - width * 3;
                        for(int i = 0; i < height; i++)
                        {
                            for(int j = 0; j < width; j++)
                            {
                                blue = *(p++);
                                avgB += blue;
                                green = *(p++);
                                avgG += green;
                                red = *(p++);
                                avgR += red;

                                px[nPixel++] = ALPHA | red << 16 | green << 8 | blue;
                            }
                            p += nOffset;
                        }
                    }
                }
            }

            map.UnlockBits(bmData);
            if(disposemap)
                map.Dispose();

            ulong np = (ulong)(width * height);

            blue = (byte)(avgB / np);
            green = (byte)(avgG / np);
            red = (byte)(avgR / np);
            if(hasAlpha)
            {
                alpha= (byte)(avgA / np);
                averageColor = alpha << 24 | red << 16 | green << 8 | blue;
            }
            else
                averageColor = ALPHA | red << 16 | green << 8 | blue;

            maxmips = bitHeight;
            if(maxmips > bitWidth)
                maxmips = bitWidth;

            maxmips -= 3; // less 1x1 2x2 and max
            hasmips = false;
        }

        public void GenMips()
        {
            int nmips = maxmips + 1;
            mips = new int[nmips + 1][];
            mips[nmips] = pixel;

            mipsBitWidth = new int[maxmips + 1];
            mipsBitHeight = new int[maxmips + 1];
            mipstw = new int[maxmips + 1];
            mipsth = new int[maxmips + 1];
  
            int w = width;
            int h = height;
            int bw = bitWidth;
            int bh = bitHeight;

            int[] src;
            int[] dst = null;
            for(int n = maxmips; n >= 0; --n)
            {
                src = mips[n + 1];
                if(w > 1)
                {
                    w >>= 1;
                    --bw;
                    if(h > 1)
                    {
                        h >>= 1;
                        --bh;
                        dst = dec2XY(src, w, h);
                    }
                    else
                        dst = dec2X(src, w, h);
                }
                else
                {
                    if(h > 1)
                    {
                        h >>= 1;
                        --bh;
                        dst = dec2Y(src, w, h);
                    }
                }
                mips[n] = dst;
                mipsBitWidth[n] = bw;
                mipsBitHeight[n] = bh;
                mipstw[n] = w;
                mipsth[n] = h;
            }
            hasmips = true;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        private unsafe int[] dec2XY(int[] src, int w, int h)
        {
            int[] dst = new int[w * h];
            int sw = 2 * w;
            int stot = 2 * h * sw;
            int c1;
            int c2;
            int c3;
            int c4;
            int sxl2 = sw;
            int dx = 0;
            fixed (int* s = src, d = dst)
            {
                for (int sx = 0 ; sx < stot; sx += sw, sxl2 += sw)
                {
                    for(int i = 0; i < w ; i++, sx += 2, sxl2 += 2)
                    {
                        c1 = s[sx];
                        c2 = s[sx + 1];
                        c3 = s[sxl2];
                        c4 = s[sxl2 + 1];
                        d[dx++] = warp_Color.avg4c(c1, c2, c3, c4);
                    }
                }
            }
            return dst;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        private int[] dec2X(int[] src, int w, int h)
        {
            int dstsize = w * h;
            int[] dst = new int[dstsize];
            int c1;
            int c2;
            int sx = 0;

            for(int dx = 0 ; dx < dstsize; ++dx)
            {
                c1 = src[sx];
                c2 = src[sx + 1];

                dst[dx] = warp_Color.avg2c(c1, c2);
                sx += 2;
            }

            return dst;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        private unsafe int[] dec2Y(int[] src, int w, int h)
        {
            int[] dst = new int[w * h];
            int stot = 2 * h * w;
            int c1;
            int c2;
            int sxl2 = w;
            int dx = 0;
            fixed (int* s = src, d = dst)
            {
                for (int sx = 0 ; sx < stot; sx += w, sxl2 += w)
                {
                    for(int i = 0; i < w ; i++, ++sx, ++sxl2)
                    {
                        c1 = s[sx];
                        c2 = s[sxl2];
                        d[dx++] = warp_Color.avg2c(c1, c2);
                    }
                }
            }
            return dst;
        }

        private unsafe void setSize(int w, int h)
        {
            int offset = w * h;
            int offset2;
            if(w * h != 0)
            {
                int[] newpixels = new int[w * h];
                fixed(int* np = newpixels, p = pixel)
                {
                    for(int j = h - 1; j >= 0; j--)
                    {
                        offset -= w;
                        offset2 = (j * height / h) * width;
                        for(int i = w - 1; i >= 0; i--)
                            np[i + offset] = p[(i * width / w) + offset2];
                    }
                }
                width = w;
                height = h;
                pixel = newpixels;
            }
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        private bool inrange(int a, int b, int c)
        {
            return (a >= b) & (a < c);
        }

        public warp_Texture getClone()
        {
            warp_Texture t = new warp_Texture(width, height);
            warp_Math.copyBuffer(pixel, t.pixel);
            return t;
        }

        public static Bitmap ResizeBitmap(Image image, int width, int height, PixelFormat format)
        {
            Bitmap result = new Bitmap(width, height, format);

            using (ImageAttributes atrib = new ImageAttributes())
            using (Graphics graphics = Graphics.FromImage(result))
            {
                atrib.SetWrapMode(System.Drawing.Drawing2D.WrapMode.TileFlipXY);
                atrib.SetWrapMode(System.Drawing.Drawing2D.WrapMode.TileFlipXY);
                graphics.CompositingQuality = System.Drawing.Drawing2D.CompositingQuality.HighQuality;
                graphics.InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic;
                graphics.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
                graphics.PixelOffsetMode = System.Drawing.Drawing2D.PixelOffsetMode.None;

                graphics.DrawImage(image,new Rectangle(0,0, result.Width, result.Height),
                    0, 0, image.Width, image.Height, GraphicsUnit.Pixel, atrib);
            }

            return result;
        }

    }
}
